Rubberduck.Parsing/VBA/ParseCoordinator.cs

```diff
@@ -69,12 +75,34 @@ public class ParseCoordinator : IParseCoordinator
 
         private readonly object _cancellationSyncObject = new object();
         private readonly object _parsingRunSyncObject = new object();
+        private readonly ReaderWriterLockSlim _parsingSuspendLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
 
         private void ReparseRequested(object sender, EventArgs e)
         {
             CancellationToken token;
             lock (_cancellationSyncObject)
             {
+                try
+                {
+                    if (!_parsingSuspendLock.IsWriteLockHeld)
+                    {
+                        _parsingSuspendLock.EnterReadLock();
```

Comment:
That is correct. The `if` statement primarily exists to avoid the need for recursion policy to allow recursions because if the thread running the `SuspendRequested` calls the reparse as a side effect of executing the `e.BusyAction.Invoke()`, we avoid the recursion exceptions in that special case. For any other threads, they will block, as you said. I think I will add a xml-doc here to clarify the intention of the `if` statement there. 

Writing it out, I realize that if the `BusyAction` delegate is allowed to start new asynchronous tasks that requires reparse, it may deadlock or at least hang until the thread executing the `BusyAction` has finished and thus exited the write lock. I will add another xml doc to discourage multithreading for the `BusyAction` to avoid those effects. For the test engine (the current sole user), we raise some events which I think should be OK.

Reply:
I made this comment because this completely invalidates the concept of the requestor stack. All new requests will block before reaching the corresponding section. My impression was that the goal of the requestor stack was to not block the threads of further parse requests.

Reply:
Btw, all parses are performed in a new task. So, the chance to have a parse reenter on the same thread is rather small.

Reply:
Crud. You're right. My intention there was to ensure that when I do `Interlocked.Read`, it wouldn't be invalid when I later execute `_requestorStack.Push(sender);`. As you see, the read lock is held only for that duration; the `BeginParse` is outside that lock. 

However, any threads that subsequently request reparse definitely shouldn't have to block; only push to the stack. 

I may make the write lock scope smaller; not holding it while `e.BusyAction.Invoke()` is being executed and ask for write lock again when popping the stack & updating the `_suspensionIteration`.

Reply:
If you do not hold the write lock during the the busy action, it can happen that a parse gets stared in parallel by a thread that had been in `Cancel` at the time the write lock was taken.